### ISSUE TITLE
ConditionBuilder incorrectly caching results when there is a hashcode collision between two objects being validated

### ISSUE THREAD

**Issue by extremeandy** [FluentValidation/FluentValidation#1810]
### FluentValidation version

10.30

### ASP.NET version

.NET 5

### Summary

When validating a collection of items, if there are different types of items in the collection which have a hashcode collision, the result of calling `ConditionBuilder.Condtiion()` is cached incorrectly. This means the caller will attempt to invoke a validator of the wrong type in some scenarios. This results in an `InvalidCastException` being thrown as it attempts to apply a rule of the wrong type to an item.

This is probably due to the following code, which seems to assume it is possibly to create a unique cache key using a hashcode: `cacheId = id + context.InstanceToValidate.GetHashCode();`.

### Steps to Reproduce

The following contrived example can be used to reproduce the issue.

```csharp
void Main() {
    var derived1 = new Derived1();
    var derived2 = new Derived2();

    var container = new Container {
        Items = new BaseClass[] { derived1, derived2 }
    };

    var validator = new ContainerValidator();
    validator.Validate(container); // Throws an InvalidCastException :(
}

abstract class BaseClass {
}

class Derived1 : BaseClass {
    public override int GetHashCode() => 1; // Hashcode collision with Derived2!
}

class Derived2 : BaseClass {
    public override int GetHashCode() => 1; // Hashcode collision with Derived1!
}

class Container {
    public BaseClass[] Items { get; set; }
}

class ContainerValidator : AbstractValidator<Container> {
    public ContainerValidator() {
        RuleForEach(c => c.Items)
            .SetValidator(new BaseClassValidator());
    }
}

class BaseClassValidator : AbstractValidator<BaseClass> {
    public BaseClassValidator() {
        When(
            model => model is Derived1,
            () => {
                RuleFor(model => ((Derived1)model))
                    .SetValidator(new Derived1Validator());
            });

        When(
            model => model is Derived2,
            () => {
                RuleFor(model => ((Derived2)model))
                    .SetValidator(new Derived2Validator());
            });
    }
}

class Derived1Validator : AbstractValidator<Derived1> {
}

class Derived2Validator : AbstractValidator<Derived2> {
}
```

**Comment by JeremySkinner** [comment]
Yep, that'll happen if 2 types are returning the same hashcode. You shouldn't really use When to do this - I'd suggest you use inheritance validation instead, which is designed for this scenario: https://docs.fluentvalidation.net/en/latest/inheritance.html?highlight=inheritance

```csharp
RuleFor(model => model).SetInheritanceValidator(v => {
  v.Add<Derived1>(new Derived1Validator());
  v.Add<Derived2>(new Derived2Validator());
});
````

**Comment by extremeandy** [comment]
That's great, thankyou. I had previously been using the `PolymorphicValidator` pattern (thought I omitted it from the above steps to reproduce), but this solution is much cleaner.

**Comment by JeremySkinner** [comment]
The hashcode issue should be fixed by #1811. This will go in the next release.

**Comment by JeremySkinner** [comment]
10.3.1 has been released
